### ISSUE TITLE
escaping \l

### ISSUE THREAD

**Issue by senya** [xflr6/graphviz#82]
Hi!

Looks like a degradation in https://github.com/xflr6/graphviz/commit/a29eaa0bb52fb7a1ffeb8ad6cb05b07b059a064d
My script uses '\l' in lables [it's DOT syntax], and now it is broken, as backslash is escaped itself.
[script is here: https://git.qemu.org/?p=qemu.git;a=blob_plain;f=scripts/render_block_graph.py;hb=HEAD;h=656f0388ad6b05840f1a77b1e150284c5ae89736;hb=HEAD ]

**Comment by xflr6** [comment]
Thanks, see #80. The specification does not seem to be clear to mere on this point.
- relevant section for these escapes: https://www.graphviz.org/doc/info/attrs.html#k:escString
- script:  https://git.qemu.org/?p=qemu.git;a=blob_plain;f=scripts/render_block_graph.py;hb=HEAD;h=656f0388ad6b05840f1a77b1e150284c5ae89736;hb=HEAD

Does using a regular `\n` instead of `\l` to separate lines help?

**Comment by xflr6** [comment]
@senya: can you try if 9d0ec14b485f7b918cc2e32d2862960b01dc5aa2 would suffice (I have never used any  of `\\G`, `\\N`, `\\E`, `\\T`, `\\H`, `\\L`, `\\n`, `\\l`, `\\r` so far)?

Not sure if this complicates things too much, e.g. `label='\\\\n'` for a literal backslash + n. An alternative might be to only backslash-escape the double quote (i.e. revert a29eaa0bb52fb7a1ffeb8ad6cb05b07b059a064d), so that doubling of backslashes for a literal backslash would be left to the user (see #80). @DNGros,  WDYT?

**Comment by DNGros** [comment]
Sorry! I was not aware of these additional escape sequences.

9d0ec14 is looks like a good covering of https://www.graphviz.org/doc/info/attrs.html#k:escString . Just reading the doc without experimenting in graphviz, it looks like the use of `\n`, `\l` and `\r` might only apply inside label, headlabel, and taillabel, while 9d0ec14 might escape for all attributes. This is somewhat obscure point though, and might need to experiment some to see how it is actually implemented.

However, my thought is that it might perhaps be best to have a settable option on whether the strings should be treated as "raw"/"render as given" and "not-raw(?)"/"escape it myself" mode, as they represent two different use cases.

In my use case my use case I was rendering user-defined strings loaded from an external file as the node labels, and expected the label to render exactly as the stored string, without the possibility of a user-given "injection" that might mess with the rendering.

However, in the use case where the developer is specifying the labels in their own code (which I would guess is most use cases?), they would likely realize they might need to escape it themselves and might want access to these special DOT escape sequences.

Not sure where would be the best place for this API to be toggled. Perhaps as "_raw" kwarg on Dot.node and Dot.edge? Alternatively maybe as "treat_strings_as_raw" arg in the Dot constructor?

For backwards-compatibility / DOT-escape-sequence-compatibility it likely might make sense to default to `_raw=False` mode. This is just an extra nice-to-have, but ideally the interface would "fail fast" if given an unrenderable string during declaration, such as raising an exception if given something like a trailing backslash when `_raw=False` and letting the user know that they either need to escape the backslash or set `_raw=True`. This would be better than waiting until the user tries to render it only to fail.

I've only been using graphviz for a short period of time, so not sure how these escaping issues are handled in other libraries in the ecosystem.

Sorry again for the regression. Let me know where I can help.

**Comment by xflr6** [comment]
No worries at all, same for me. :) +1 that there are these two use cases. So far I tried to avoid adding such a `raw=True` parameter to all the methods that accept strings. I suspect that some users would prefer to have easy access to the escapes, yet that might be a small number of overall users.

At the moment I am thinking of it to work like the build-in `re` module: users need to escape their escapes (use raw strings in most cases) and whenever they want to pass user-data, use `re.escape` for sanitization first (`graphviz.escape()` would be added for that).

Prototype in cc03918f755edde93febb03df38fffbde420c632 ([nbviewer](https://nbviewer.jupyter.org/github/xflr6/graphviz/blob/escString/graphviz-escaping.ipynb)), let me know what you think (note that unescaped quotes are still escaped, so the escaping should be idempotent now).

**Comment by wataash** [comment]
FWIW: in my use case, I've used `r`-prefixed strings with `\n` for labels, and today the same issue hit me.

```py
import graphviz

g = graphviz.Digraph()
g.node('foo', label=r'high\nthere')
g.view()
```

**Comment by xflr6** [comment]
Thanks @wataash. Can you maybe take a look at the pending solution in 041fa44962390d0eebd39d9359175fbb03ced74e ([nbviewer with some usage examples](https://nbviewer.jupyter.org/github/xflr6/graphviz/blob/escString/examples/graphviz-escapes.ipynb)) and tell me if that suffices for your use case?

**Comment by wataash** [comment]
@xflr6 It resolved the regression, thanks!

**Comment by wataash** [comment]
For those who reached here via google:
your problem might be resolved by: `pip install git+https://github.com/xflr6/graphviz@escString`

**Comment by xflr6** [comment]
Thanks. Note that `escString` is just a dev branch: will merge and create a new release.

**Comment by xflr6** [comment]
Fixed in [0.13 release](https://pypi.org/project/graphviz/0.13/#files) (84abaff30f6ca3f4c5c152691002720a319dc876)
